### PR TITLE
[HOPSWORKS-3202] Add filter to td

### PR DESCRIPTION
**Pull request by kennethmhc**
changes:
1. Fix query parameter in feature view API
2. Add extra filter to all get/create TD
3. Add td version to get batch query/data.
4. Add extra filter to td entity.
5. Remove default td version = 1 when init batch scoring since backend will not be able to fetch if there is none.

### CODE REVIEW

File: java/src/main/java/com/logicalclocks/hsfs/FeatureView.java [hunk 1]
```diff
@@ -183,31 +185,34 @@ public List<List<Object>> previewFeatureVectors(Integer n, boolean external)
...
   @JsonIgnore
-  public Dataset<Row> getBatchData(String startTime, String endTime)
+  public Dataset<Row> getBatchData(String startTime, String endTime, Integer trainingDataVersion)
       throws FeatureStoreException, IOException, ParseException {
-    return getBatchData(startTime, endTime, Maps.newHashMap());
+    return getBatchData(startTime, endTime, Maps.newHashMap(), trainingDataVersion);
   }
 
-  public Dataset<Row> getBatchData(String startTime, String endTime, Map<String, String> readOptions)
+  public Dataset<Row> getBatchData(String startTime, String endTime, Map<String, String> readOptions,
```
Comment (jimdowling) [inline java/src/main/java/com/logicalclocks/hsfs/FeatureView.java:207]: One thing about filters for training data is that they should only be applied when creating the training data. 
When i get batch data for scoring, often i don't want the filters to be applied.
If I don't supply the TD_Version here, will I get all the batch data in the time interval?

Comment (jimdowling) [reply java/src/main/java/com/logicalclocks/hsfs/FeatureView.java:207]: I have the surfing example, where i can walk through the use case with you to explain why i only want the filter on the TD data, not on the batch data for scoring.

File: python/hsfs/feature_view.py [hunk 2]
```diff
@@ -161,7 +154,14 @@ def get_batch_query(
             `str`: batch query
         """
         return self._feature_view_engine.get_batch_query_string(
-            self, start_time, end_time
+            self,
+            start_time,
+            end_time,
+            training_dataset_version=(
```
Comment (jimdowling) [inline python/hsfs/feature_view.py:160]: Again, the TD_version should be an optional parameter for get_batch_data.

File: java/src/main/java/com/logicalclocks/hsfs/metadata/FeatureViewApi.java [hunk 3]
```diff
@@ -292,15 +292,17 @@ public void deleteTrainingDatasetOnly(FeatureStore featureStore, String name, In
   }
 
   public Query getBatchQuery(FeatureStore featureStore, String name, Integer version,
-      Long startTime, Long endTime, Boolean withLabels) throws FeatureStoreException, IOException {
+      Long startTime, Long endTime, Boolean withLabels, Integer trainingDataVersion) throws FeatureStoreException,
```
Comment (jimdowling) [inline java/src/main/java/com/logicalclocks/hsfs/metadata/FeatureViewApi.java:295]: Optional TD version.

**Review by jimdowling — CHANGES_REQUESTED** [review]
I mentioned in the comments, for get_batch_data(), it should be optional to supply the TD_Version. In the only use case I have, i don't want to supply the TD_version on get_batch_data() - the surfing example. In the other theoretical examples we had before, such as "models for users from US/EU/Asia", we might want to batch score separately, and therefore apply the TD_version in get_batch_data(). I suspect we need both - where I would have it off by default and only if user supplies the TD_version, apply it.

File: java/src/main/java/com/logicalclocks/hsfs/FeatureView.java (hunk 1 above)
Comment (kennethmhc) [reply java/src/main/java/com/logicalclocks/hsfs/FeatureView.java:207]: > If I don't supply the TD_Version here, will I get all the batch data in the time interval?
yes, td version is optional in the current setting, you can pass `null` here.
> explain why i only want the filter on the TD data, not on the batch data for scoring.
To do this, we need to introduce a new flag "ignore_td_filter" because by default, when td contains statistics as well. If you do not provide td version, no transformation will occur. But I think having the extra may not be necessary. In your case, you can create a new td without filter, and get batch data from there.

Comment (jimdowling) [reply java/src/main/java/com/logicalclocks/hsfs/FeatureView.java:207]: Good point!
We would still want to apply the transformations and use the TD_version statistics. 
So a new flag - ignore_td_filter, would do the job. Is there a limit to 1 filter or can you add many filters? (ignore_td_filters)

We talked at the feature monitoring meeting last week about having a separate way to log the untransformed features - i know it's not directly related, but you might want to add an explicit flag for that - ignore_transformations.

Comment (jimdowling) [reply java/src/main/java/com/logicalclocks/hsfs/FeatureView.java:207]: Passing null as a parameter doesn't feel right though. Can we make it an optional parameter in Python (In java, it's not as big a deal, IMO).

File: java/src/main/java/com/logicalclocks/hsfs/FeatureView.java [hunk 4]
```diff
@@ -183,31 +185,34 @@ public List<List<Object>> previewFeatureVectors(Integer n, boolean external)
 
   @JsonIgnore
   public String getBatchQuery() throws FeatureStoreException, IOException, ParseException {
-    return getBatchQuery(null, null);
+    return getBatchQuery(null, null, null);
   }
 
   @JsonIgnore
-  public String getBatchQuery(String startTime, String endTime)
+  public String getBatchQuery(String startTime, String endTime, Integer trainingDataVersion)
```
Comment (bubriks) [inline java/src/main/java/com/logicalclocks/hsfs/FeatureView.java:192]: maybe a new method could be added that does not require `trainingDataVersion` (in a sense proxy method). This could be applicable in other places as well.

File: java/src/main/java/com/logicalclocks/hsfs/engine/FeatureViewEngine.java [hunk 5]
```diff
@@ -170,43 +167,44 @@ public TrainingDatasetBundle getTrainingDataset(
...
-            featureView.getLabels());
-        computeStatistics(featureView, trainingDatasetUpdated, datasets);
-      } else {
-        trainingDatasetBundle = new TrainingDatasetBundle(trainingDatasetUpdated.getVersion(), dataset,
-            featureView.getLabels());
-        computeStatistics(featureView, trainingDatasetUpdated, new Dataset[]{dataset});
-      }
-      return trainingDatasetBundle;
-    }
+//    if (!IN_MEMORY_TRAINING_DATASET.equals(trainingDatasetUpdated.getTrainingDatasetType())) {
```
Comment (bubriks) [inline java/src/main/java/com/logicalclocks/hsfs/engine/FeatureViewEngine.java:170]: why did it have to be commented out? why not just remove it if it is not necessary?

**Review by bubriks — COMMENTED** [review]
overall looks good, just minor comments/questions

File: python/hsfs/feature_view.py [hunk 6]
```diff
@@ -161,7 +154,14 @@ def get_batch_query(
             `str`: batch query
         """
         return self._feature_view_engine.get_batch_query_string(
-            self, start_time, end_time
+            self,
```
Comment (tdoehmen) [inline python/hsfs/feature_view.py:157]: New parameter still needs to be added to the docstring,

**Review by tdoehmen — COMMENTED** [review]
Merge conflicts with master still need to be solved

File: python/hsfs/core/statistics_engine.py [hunk 7]
```diff
@@ -78,7 +78,7 @@ def profile_statistics(metadata_instance, feature_dataframe):
                 "to the online storage of a feature group.",
                 category=util.StatisticsWarning,
             )
-            return None
+            return "{}"
```
Comment (tdoehmen) [inline python/hsfs/core/statistics_engine.py:81]: Is this change related to the PR`?

File: python/hsfs/feature_view.py (hunk 6 above)
Comment (kennethmhc) [reply python/hsfs/feature_view.py:157]: Do you mean td version? It is not required in the parameter.

File: python/hsfs/core/statistics_engine.py (hunk 7 above)
Comment (kennethmhc) [reply python/hsfs/core/statistics_engine.py:81]: I had an error when testing this PR. If the dataframe is empty, it causes error in the backend.

File: java/src/main/java/com/logicalclocks/hsfs/FeatureView.java [hunk 8]
```diff
@@ -162,7 +164,7 @@ public List<List<Object>> getFeatureVectors(Map<String, List<Object>> entry, boo
 
   @JsonIgnore
   public String getBatchQuery() throws FeatureStoreException, IOException, ParseException {
-    return getBatchQuery(null, null);
+    return getBatchQuery(null, null, null);
```
Comment (bubriks) [inline java/src/main/java/com/logicalclocks/hsfs/FeatureView.java:167]: no need for this change

File: java/src/main/java/com/logicalclocks/hsfs/FeatureView.java [hunk 9]
```diff
@@ -171,22 +173,41 @@ public String getBatchQuery(String startTime, String endTime)
     return featureViewEngine.getBatchQueryString(
```
Comment (bubriks) [inline java/src/main/java/com/logicalclocks/hsfs/FeatureView.java:173]: instead of calling `featureViewEngine.getBatchQueryString` could be better to call `this.getBatchQuery` to avoid duplication.

File: java/src/main/java/com/logicalclocks/hsfs/FeatureView.java [hunk 10]
```diff
@@ -382,14 +410,17 @@ public Integer createTrainValidationTestSplit(
...
+            .extraFilter(extraFilter)
             .build();
     return featureViewEngine.createTrainingDataset(this, trainingDataset, writeOptions).getVersion();
   }
 
   private List<Dataset<Row>> getDataset(TrainingDatasetBundle trainingDatasetBundle, List<String> splits) {
-    return splits.stream()
-        .flatMap(split -> trainingDatasetBundle.getDataset(split, true).stream())
-        .collect(Collectors.toList());
+    return null;
```
Comment (bubriks) [inline java/src/main/java/com/logicalclocks/hsfs/FeatureView.java:420]: why was this replaced?

File: java/src/main/java/com/logicalclocks/hsfs/constructor/Filter.java [hunk 11]
```diff
@@ -19,8 +19,9 @@
 import com.logicalclocks.hsfs.Feature;
 import lombok.AllArgsConstructor;
 import lombok.Getter;
+import lombok.NoArgsConstructor;
 import lombok.Setter;
-
+@NoArgsConstructor
```
[the remark this answers is not in the document]
Comment (bubriks) [reply java/src/main/java/com/logicalclocks/hsfs/constructor/Filter.java:24]: this issue looks to persist

**Review by bubriks — COMMENTED** [review]
minor comments

**Comment by bubriks** [comment]
also python lint fails

**Review by bubriks — APPROVED** [review]
Looks good to me.

**Comment by tdoehmen** [comment]
Checkstyle fails in java client

[ERROR] C:\Users\doehmen\Documents\Hopsworks\feature-store-api\java\src\main\java\com\logicalclocks\hsfs\engine\FeatureViewEngine.java:238:73: '&&' should be on a new line. [OperatorWrap]
[ERROR] C:\Users\doehmen\Documents\Hopsworks\feature-store-api\java\src\main\java\com\logicalclocks\hsfs\engine\FeatureViewEngine.java:239:51: '&&' should be on a new line. [OperatorWrap]
[ERROR] C:\Users\doehmen\Documents\Hopsworks\feature-store-api\java\src\main\java\com\logicalclocks\hsfs\engine\FeatureViewEngine.java:243:73: '&&' should be on a new line. [OperatorWrap]
[ERROR] C:\Users\doehmen\Documents\Hopsworks\feature-store-api\java\src\main\java\com\logicalclocks\hsfs\engine\FeatureViewEngine.java:244:50: '&&' should be on a new line. [OperatorWrap]
Audit done.

File: python/hsfs/feature_view.py (hunk 6 above)
Comment (tdoehmen) [reply python/hsfs/feature_view.py:157]: I was anticipating that you are going to add the td version parameter to that function, based on Jim's comment and the fact that the Java API getBatchQuery method has that version parameter. I don't know what the conclusion was between you and Jim was, but I think the API should be consistent between Java and Python.

File: python/hsfs/core/statistics_engine.py (hunk 7 above)
Comment (tdoehmen) [reply python/hsfs/core/statistics_engine.py:81]: Was this an interegration test that failed?

**Review by jimdowling — APPROVED** [review]

File: python/hsfs/training_dataset.py [hunk 12]
```diff
@@ -157,6 +161,8 @@ def __init__(
                 else extra_filter
             )
         else:
+            self._start_time = event_start_time
```
Comment (kennethmhc) [inline python/hsfs/training_dataset.py:160]: `event_start_time` and `event_end_time` do not need to transformed if they are returned from backend.

File: python/hsfs/engine/spark.py [hunk 13]
```diff
@@ -854,14 +854,14 @@ def save_empty_dataframe(self, feature_group, dataframe):
             {},
         )
 
-    def _apply_transformation_function(self, training_dataset, dataset):
+    def _apply_transformation_function(self, transformation_functions, dataset):
```
Comment (kennethmhc) [inline python/hsfs/engine/spark.py:857]: only transformation functions is sufficient here.

File: python/hsfs/feature_view.py [hunk 14]
```diff
@@ -132,7 +132,7 @@ def init_batch_scoring(
         """Initialise and cache parametrized transformation functions.
 
         # Arguments
-            training_dataset_version: int, optional. Default to be 1. Transformation statistics
+            training_dataset_version: int, optional. Default to be None. Transformation statistics
```
Comment (kennethmhc) [inline python/hsfs/feature_view.py:135]: TD version cannot be default to 1 here because the backend will try to retrieve the td metadata and get the extra filter. If we set the default to 1, it will cause 404 error. @davitbzh any comments?

**Comment by tdoehmen** [comment]
When passing the extra filter in the Java API, the server complains:

`java.io.IOException: Error: 422{"type":"restApiJsonResponse","errorCode":120001,"errorMsg":"An argument was not provided or it was malformed.","usrMsg":"2022-01-06"}`

````
val filter = fv.getQuery().getLeftFeatures().get(0).gt("2022-01-06")
val ret = fv.createTrainingData(null, null, "transactions fraud batch training dataset", DataFormat.CSV,  true, null,null,null,null,null,null,filter);

**Comment by tdoehmen** [comment]
All looks good. As soon as the tests passt I think we can merge

**Comment by kennethmhc** [comment]
@tdoehmen I have fixed the tests.
